File: greeting-card-web/services/cart.service.ts
import { apiClient } from './client';
import type {
  AddToCartRequest,
  ApiResponse,
  CartResponse,
  CartSummary,
  PaginationParams,
  ServiceResponse,
  SyncCartRequest,
  UpdateCartItemRequest,
} from '@/types';

// Lấy giỏ hàng hiện tại từ server
export const getCart = async (): Promise<ServiceResponse<CartResponse>> => {
  const response = await apiClient.get<ApiResponse<CartResponse>>('/cart');
  return {
    data: response.data.data!,
    message: response.data.message,
  };
};

// Thêm sản phẩm vào giỏ hàng
export const addToCart = async (
  request: AddToCartRequest,
): Promise<ServiceResponse<CartResponse>> => {
  const response = await apiClient.post<ApiResponse<CartResponse>>('/cart/add', request);
  return {
    data: response.data.data!,
    message: response.data.message,
  };
};

// Cập nhật số lượng sản phẩm trong giỏ hàng
export const updateCartItem = async (
  productId: number,
  request: UpdateCartItemRequest,
): Promise<ServiceResponse<CartResponse>> => {
  const response = await apiClient.put<ApiResponse<CartResponse>>(
    `/cart/items/${productId}`,
    request,
  );
  return {
    data: response.data.data!,
    message: response.data.message,
  };
};

// Xóa sản phẩm khỏi giỏ hàng
export const removeCartItem = async (productId: number): Promise<ServiceResponse<void>> => {
  const response = await apiClient.delete<ApiResponse<void>>(`/cart/items/${productId}`);
  return {
    data: undefined as void,
    message: response.data.message,
  };
};

// Xóa toàn bộ giỏ hàng
export const clearCart = async (): Promise<ServiceResponse<void>> => {
  const response = await apiClient.delete<ApiResponse<void>>('/cart');
  return {
    data: undefined as void,
    message: response.data.message,
  };
};

// Sync giỏ hàng từ localStorage lên server khi user login
export const syncCart = async (
  request: SyncCartRequest,
): Promise<ServiceResponse<CartResponse>> => {
  const response = await apiClient.post<ApiResponse<CartResponse>>('/cart/sync', request);
  return {
    data: response.data.data!,
    message: response.data.message,
  };
};

// === ADMIN APIs ===

// Admin: Lấy tất cả giỏ hàng (phân trang)
export const getAllCarts = async (
  params?: PaginationParams,
): Promise<ServiceResponse<CartSummary[]>> => {
  const response = await apiClient.get<ApiResponse<CartSummary[]>>('/admin/carts', { params });
  return {
    data: response.data.data!,
    message: response.data.message,
    pagination: response.data.pagination,
  };
};

// Admin: Tìm kiếm giỏ hàng theo email hoặc tên user
export const searchCarts = async (
  keyword: string,
  params?: PaginationParams,
): Promise<ServiceResponse<CartSummary[]>> => {
  const response = await apiClient.get<ApiResponse<CartSummary[]>>('/admin/carts/search', {
    params: { keyword, ...params },
  });
  return {
    data: response.data.data!,
    message: response.data.message,
    pagination: response.data.pagination,
  };
};

// Admin: Xem chi tiết giỏ hàng theo ID
export const getCartById = async (cartId: number): Promise<ServiceResponse<CartResponse>> => {
  const response = await apiClient.get<ApiResponse<CartResponse>>(`/admin/carts/${cartId}`);
  return {
    data: response.data.data!,
    message: response.data.message,
  };
};

// Admin: Xem giỏ hàng của một user cụ thể
export const getUserCart = async (userId: number): Promise<ServiceResponse<CartResponse>> => {
  const response = await apiClient.get<ApiResponse<CartResponse>>(`/admin/carts/user/${userId}`);
  return {
    data: response.data.data!,
    message: response.data.message,
  };
};

// Admin: Xóa giỏ hàng của user (hỗ trợ khách hàng)
export const clearUserCart = async (userId: number): Promise<ServiceResponse<void>> => {
  const response = await apiClient.delete<ApiResponse<void>>(`/admin/carts/user/${userId}`);
  return {
    data: undefined as void,
    message: response.data.message,
  };
};

File: greeting-card-api/src/main/java/iuh/fit/se/service/CartService.java
package iuh.fit.se.service;

import java.math.BigDecimal;
import java.util.List;
import java.util.stream.Collectors;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import iuh.fit.se.dto.request.AddToCartRequest;
import iuh.fit.se.dto.request.SyncCartRequest;
import iuh.fit.se.dto.request.UpdateCartItemRequest;
import iuh.fit.se.dto.response.CartItemResponse;
import iuh.fit.se.dto.response.CartResponse;
import iuh.fit.se.dto.response.CartSummaryResponse;
import iuh.fit.se.dto.response.ProductImageResponse;
import iuh.fit.se.dto.response.ProductResponse;
import iuh.fit.se.entity.Cart;
import iuh.fit.se.entity.CartItem;
import iuh.fit.se.entity.Product;
import iuh.fit.se.entity.User;
import iuh.fit.se.exception.ResourceNotFoundException;
import iuh.fit.se.repository.CartItemRepository;
import iuh.fit.se.repository.CartRepository;
import iuh.fit.se.repository.ProductRepository;
import iuh.fit.se.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Service
@RequiredArgsConstructor
@Transactional
@Slf4j
public class CartService {
  private final CartRepository cartRepository;
  private final CartItemRepository cartItemRepository;
  private final ProductRepository productRepository;
  private final UserRepository userRepository;

  @Transactional
  public CartResponse getCart(Long userId) {
    Cart cart = getOrCreateCart(userId);
    return mapToCartResponse(cart);
  }

  // Lấy giỏ hàng theo cartId (dùng cho admin)
  @Transactional(readOnly = true)
  public CartResponse getCartById(Long cartId) {
    Cart cart =
        cartRepository
            .findByIdWithItems(cartId)
            .orElseThrow(() -> new ResourceNotFoundException("Giỏ hàng không tồn tại"));
    return mapToCartResponse(cart);
  }

  // Thêm sản phẩm vào giỏ hàng
  public CartResponse addToCart(Long userId, AddToCartRequest request) {
    Cart cart = getOrCreateCart(userId);
    Product product = getProductOrThrow(request.getProductId());

    // Validate stock
    if (product.getStock() < request.getQuantity()) {
      throw new IllegalArgumentException(
          "Sản phẩm không đủ số lượng trong kho. Còn lại: " + product.getStock());
    }

    // Check if product already exists in cart
    CartItem existingItem =
        cartItemRepository.findByCartIdAndProductId(cart.getId(), product.getId()).orElse(null);

    if (existingItem != null) {
      // Update quantity
      int newQuantity = existingItem.getQuantity() + request.getQuantity();
      if (product.getStock() < newQuantity) {
        throw new IllegalArgumentException(
            "Không thể thêm. Tồn kho chỉ còn: " + product.getStock());
      }
      existingItem.setQuantity(newQuantity);
      cartItemRepository.save(existingItem);
    } else {
      // Add new item
      CartItem newItem = new CartItem();
      newItem.setCart(cart);
      newItem.setProduct(product);
      newItem.setQuantity(request.getQuantity());
      cart.addItem(newItem);
      cartItemRepository.save(newItem);
    }

    log.info("Added product {} to cart for user {}", request.getProductId(), userId);

    return mapToCartResponse(cart);
  }

  // Cập nhật số lượng của item trong giỏ hàng (quantity = 0 sẽ xóa item)
  public CartResponse updateCartItem(Long userId, Long productId, UpdateCartItemRequest request) {
    Cart cart = getCartByUserIdOrThrow(userId);

    CartItem cartItem =
        cartItemRepository
            .findByCartIdAndProductId(cart.getId(), productId)
            .orElseThrow(() -> new ResourceNotFoundException("Sản phẩm không có trong giỏ hàng"));

    if (request.getQuantity() == 0) {
      // Remove item
      cart.removeItem(cartItem);
      cartItemRepository.delete(cartItem);
      log.info("Removed product {} from cart for user {}", productId, userId);
    } else {
      // Update quantity
      Product product = cartItem.getProduct();
      if (product.getStock() < request.getQuantity()) {
        throw new IllegalArgumentException(
            "Số lượng vượt quá tồn kho. Còn lại: " + product.getStock());
      }
      cartItem.setQuantity(request.getQuantity());
      cartItemRepository.save(cartItem);
      log.info(
          "Updated product {} quantity to {} in cart for user {}",
          productId,
          request.getQuantity(),
          userId);
    }

    return mapToCartResponse(cart);
  }

  // Xóa item khỏi giỏ hàng
  public void removeCartItem(Long userId, Long productId) {
    Cart cart = getCartByUserIdOrThrow(userId);

    CartItem cartItem =
        cartItemRepository
            .findByCartIdAndProductId(cart.getId(), productId)
            .orElseThrow(() -> new ResourceNotFoundException("Sản phẩm không có trong giỏ hàng"));

    cart.removeItem(cartItem);
    cartItemRepository.delete(cartItem);

    log.info("Removed product {} from cart for user {}", productId, userId);
  }

  // Xóa toàn bộ giỏ hàng
  public void clearCart(Long userId) {
    Cart cart = getCartByUserIdOrThrow(userId);
    cart.clearItems();
    cartItemRepository.deleteAll(cart.getItems());
    log.info("Cleared cart for user {}", userId);
  }

  // Sync giỏ hàng từ localStorage lên server khi user login (merge với cart hiện tại)
  public CartResponse syncCart(Long userId, SyncCartRequest request) {
    Cart cart = getOrCreateCart(userId);

    for (SyncCartRequest.CartItemData item : request.getItems()) {
      Product product = getProductOrThrow(item.getProductId());

      // Skip if out of stock
      if (product.getStock() < item.getQuantity()) {
        log.warn(
            "Skipping product {} - out of stock. Requested: {}, Available: {}",
            item.getProductId(),
            item.getQuantity(),
            product.getStock());
        continue;
      }

      CartItem existingItem =
          cartItemRepository
              .findByCartIdAndProductId(cart.getId(), item.getProductId())
              .orElse(null);

      if (existingItem != null) {
        // Merge: cộng dồn số lượng, nhưng không vượt quá stock
        int newQuantity =
            Math.min(existingItem.getQuantity() + item.getQuantity(), product.getStock());
        existingItem.setQuantity(newQuantity);
        cartItemRepository.save(existingItem);
      } else {
        // Add new item
        CartItem newItem = new CartItem();
        newItem.setCart(cart);
        newItem.setProduct(product);
        newItem.setQuantity(Math.min(item.getQuantity(), product.getStock()));
        cart.addItem(newItem);
        cartItemRepository.save(newItem);
      }
    }

    log.info("Synced cart from localStorage for user {}", userId);

    return mapToCartResponse(cart);
  }

  // === Helper Methods ===

  private Cart getOrCreateCart(Long userId) {
    return cartRepository
        .findByUserIdWithItems(userId)
        .orElseGet(
            () -> {
              User user =
                  userRepository
                      .findById(userId)
                      .orElseThrow(() -> new ResourceNotFoundException("User not found"));
              Cart newCart = new Cart();
              newCart.setUser(user);
              return cartRepository.save(newCart);
            });
  }

  private Cart getCartByUserIdOrThrow(Long userId) {
    return cartRepository
        .findByUserIdWithItems(userId)
        .orElseThrow(() -> new ResourceNotFoundException("Giỏ hàng không tồn tại"));
  }

  private Product getProductOrThrow(Long productId) {
    return productRepository
        .findById(productId)
        .orElseThrow(() -> new ResourceNotFoundException("Sản phẩm không tồn tại"));
  }

  private CartResponse mapToCartResponse(Cart cart) {
    List<CartItemResponse> items =
        cart.getItems().stream().map(this::mapToCartItemResponse).collect(Collectors.toList());

    return new CartResponse(cart.getId(), items);
  }

  private CartItemResponse mapToCartItemResponse(CartItem item) {
    Product product = item.getProduct();

    // Map product images
    List<ProductImageResponse> images =
        product.getImages().stream()
            .map(
                img ->
                    new ProductImageResponse(
                        img.getId(),
                        img.getImageUrl(),
                        img.getAltText(),
                        img.getIsPrimary(),
                        img.getDisplayOrder()))
            .collect(Collectors.toList());

    ProductResponse productResponse =
        new ProductResponse(
            product.getId(),
            product.getName(),
            product.getSlug(),
            product.getDescription(),
            product.getPrice(),
            product.getStock(),
            product.getSku(),
            product.getIsActive(),
            product.getIsFeatured(),
            null, // category
            images, // images
            null, // tags
            product.getCreatedAt(),
            product.getUpdatedAt(),
            null, // inWishlist - not needed in cart context
            true // inCart - item is in cart since it's a cart item
            );

    return new CartItemResponse(item.getId(), productResponse, item.getQuantity());
  }

  // Admin: Lấy tất cả giỏ hàng (phân trang)
  @Transactional(readOnly = true)
  public Page<CartSummaryResponse> getAllCarts(Pageable pageable) {
    Page<Cart> carts = cartRepository.findAllWithUserAndItems(pageable);
    return carts.map(this::mapToCartSummary);
  }

  // Admin: Tìm kiếm giỏ hàng theo email hoặc tên user
  @Transactional(readOnly = true)
  public Page<CartSummaryResponse> searchCarts(String keyword, Pageable pageable) {
    Page<Cart> carts = cartRepository.searchCarts(keyword, pageable);
    return carts.map(this::mapToCartSummary);
  }

  private CartSummaryResponse mapToCartSummary(Cart cart) {
    int totalItems = cart.getItems().stream().mapToInt(CartItem::getQuantity).sum();

    BigDecimal totalAmount =
        cart.getItems().stream()
            .map(
                item ->
                    item.getProduct().getPrice().multiply(BigDecimal.valueOf(item.getQuantity())))
            .reduce(BigDecimal.ZERO, BigDecimal::add);

    return CartSummaryResponse.builder()
        .id(cart.getId())
        .userId(cart.getUser().getId())
        .userEmail(cart.getUser().getEmail())
        .userFullName(cart.getUser().getFullName())
        .totalItems(totalItems)
        .totalAmount(totalAmount)
        .updatedAt(cart.getUpdatedAt())
        .build();
  }
}
